### PR TITLE
MicrosoftDI Services Auto Registration

### LINKED ISSUE #13: Implement steps DI registration using reflection.
Motivation:

The current implementation requires users to manually register all steps to DI. It can lead to writing a significant amount of useless code and this can lead to users simply forgetting to register some steps which in turn can lead to runtime errors.

### PR DESCRIPTION
**Pull request by homolibere**
Closes #13

### CODE REVIEW

File: src/PowerPipe.Extensions.MicrosoftDependencyInjection/ServiceRegistrar.cs [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+        foreach (var inf in interfaces)
+        {
+            var exactMatches = concretions.Where(x => x.CanBeCastTo(inf)).ToList();
+        
+            if (exactMatches.Count > 1)
+            {
+                exactMatches.RemoveAll(m => !IsMatchingWithInterface(m, inf));
+            }
+
+            foreach (var type in exactMatches.Where(type => !configuration.BehaviorsToRegister.Any(c => c.ImplementationType == type)))
```
Comment (mvSapphire) [inline src/PowerPipe.Extensions.MicrosoftDependencyInjection/ServiceRegistrar.cs:94]: `!configuration.BehaviorsToRegister.Any(c => c.ImplementationType == type)` -> `configuration.BehaviorsToRegister.All(c => c.ImplementationType != type)`

File: tests/PowerPipe.UnitTests/Steps/TestParallelStep.cs [hunk 2]
```diff
@@ -4,12 +4,14 @@
 
 namespace PowerPipe.UnitTests.Steps;
 
-public class TestParallelStep : IPipelineParallelStep<TestPipelineContext>
+public class TestParallelStep : IPipelineStep<TestPipelineContext>, IPipelineParallelStep<TestPipelineContext>
```
Comment (mvSapphire) [inline tests/PowerPipe.UnitTests/Steps/TestParallelStep.cs:7]: why you added implementation of `IPipelineStep<TestPipelineContext>`?

File: src/PowerPipe.Extensions.MicrosoftDependencyInjection/ServiceRegistrar.cs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        Type openRequestInterface,
+        IServiceCollection services,
+        IEnumerable<Assembly> assembliesToScan,
+        PowerPipeConfiguration configuration)
+    {
+        var concretions = new List<Type>();
+        var interfaces = new List<Type>();
+        foreach (var type in 
+                 assembliesToScan.SelectMany(a => a.DefinedTypes)
+                     .Where(t => !t.IsOpenGeneric()).Where(configuration.TypeEvaluator))
```
[the remark this answers is not in the document]
Comment (homolibere) [reply src/PowerPipe.Extensions.MicrosoftDependencyInjection/ServiceRegistrar.cs]: fixed, now we can register generic steps

File: src/PowerPipe.Extensions.MicrosoftDependencyInjection/ServiceRegistrar.cs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+using PowerPipe.Factories;
+using PowerPipe.Interfaces;
+
+namespace PowerPipe.Extensions.MicrosoftDependencyInjection;
+
+public static class ServiceRegistrar
+{
+    public static void AddPowerPipeClasses(IServiceCollection services, PowerPipeConfiguration configuration)
+    {
+        var assembliesToScan = configuration.AssembliesToRegister.Distinct().ToArray();
```
[the remark this answers is not in the document]
Comment (homolibere) [reply src/PowerPipe.Extensions.MicrosoftDependencyInjection/ServiceRegistrar.cs:15]: i believe yes

**Review by mvSapphire — APPROVED** [review]

**Pull request merged**